hydroflow_lang/src/graph/ops/join_fused.rs

```diff
@@ -0,0 +1,333 @@
+
+        let write_prologue = quote_spanned! {op_span=>
+            #lhs_prologue
+            #rhs_prologue
+        };
+
+        let lhs = &inputs[0];
+        let rhs = &inputs[1];
+
+        let arg0_span = arguments[0].span();
+        let arg1_span = arguments[1].span();
+
+        let lhs_tokens = match lhs_join_options {
+            JoinOptions::FoldFrom(lhs_from, lhs_fold) => quote_spanned! {arg0_span=>
+                #lhs_borrow.fold_into(#lhs, #lhs_fold, #lhs_from);
+            },
+            JoinOptions::Fold(lhs_default, lhs_fold) => quote_spanned! {arg0_span=>
+                #lhs_borrow.fold_into(#lhs, #lhs_fold, #lhs_default);
+            },
+            JoinOptions::Reduce(lhs_reduce) => quote_spanned! {arg0_span=>
+                #lhs_borrow.reduce_into(#lhs, #lhs_reduce);
+            },
+        };
+
+        let rhs_tokens = match rhs_join_options {
+            JoinOptions::FoldFrom(rhs_from, rhs_fold) => quote_spanned! {arg0_span=>
+                #rhs_borrow.fold_into(#rhs, #rhs_fold, #rhs_from);
+            },
+            JoinOptions::Fold(rhs_default, rhs_fold) => quote_spanned! {arg1_span=>
+                #rhs_borrow.fold_into(#rhs, #rhs_fold, #rhs_default);
+            },
+            JoinOptions::Reduce(rhs_reduce) => quote_spanned! {arg1_span=>
+                #rhs_borrow.reduce_into(#rhs, #rhs_reduce);
+            },
+        };
+
+        // Since both input arguments are stratum blocking then we don't need to keep track of ticks to avoid emitting the same thing twice in the same tick.
+        let write_iterator = quote_spanned! {op_span=>
+            let mut #lhs_borrow_ident = #context.state_ref(#lhs_joindata_ident).borrow_mut();
+            let mut #rhs_borrow_ident = #context.state_ref(#rhs_joindata_ident).borrow_mut();
+
+            let #ident = {
+                #lhs_tokens
+                #rhs_tokens
+
+                // TODO: start the iterator with the smallest len() table rather than always picking rhs.
+                #[allow(clippy::clone_on_copy)]
+                #[allow(suspicious_double_ref_op)]
+                #rhs_borrow
+                    .table
+                    .iter()
+                    .filter_map(|(k, v2)| #lhs_borrow.table.get(k).map(|v1| (k.clone(), (v1.clone(), v2.clone()))))
+            };
+        };
+
+        let write_iterator_after =
+            if persistences[0] == Persistence::Static || persistences[1] == Persistence::Static {
+                quote_spanned! {op_span=>
+                    // TODO: Probably only need to schedule if #*_borrow.len() > 0?
+                    #context.schedule_subgraph(#context.current_subgraph(), false);
+                }
+            } else {
+                quote_spanned! {op_span=>}
+            };
+
+        Ok(OperatorWriteOutput {
+            write_prologue,
+            write_iterator,
+            write_iterator_after,
+        })
+    },
+};
+
+pub enum JoinOptions<'a> {
```

Comment:
Move earlier for readability? Maybe also make the doc comment more specifically mention that the `JoinOptions` for each side are an `enum` (not "like" one, it *is* one, and it's here in the code!).

Reply:
This is actually a different enum than the one users specify when calling `join_fused()`. Technically the way it is parsed in the `join_fused()`  surface syntax is that the `Fold(<>, <>)` thing is actually a syntactic function call, where the function name is 'Fold' and the arguments go in the parentheses.

But otherwise yeah their config maps more or less to this actual enum. I removed the part about 'being like an enum' though because it's confusing when you can just show syntactic examples.

Reply:
As for moving earlier for more readability, I don't want to do that because there's a convention where the operator config is the first thing in these operator level files.

I would be open to the idea of moving these helper functions and types up to mod.rs, I'll see what @MingweiSamuel  thinks

Reply:
fine to keep them here, in the future they'll only be used in one op. also make them `pub(crate)` instead of `pub`